lib/redis_client.rb

```diff
@@ -79,10 +79,27 @@ def timeout=(timeout)
 
   Error = Class.new(StandardError)
 
-  ProtocolError = Class.new(Error)
-  UnsupportedServer = Class.new(Error)
+  class ErrorWithContext < Error
```

Comment:
Not sure why this is a class. Could be a module like `HasCommand` or just straight inside `Error`.

Reply:
The reasoning is that it will never be instantiated, so it shouldn't be a class.

Reply:
I originally tried to use a module `HasContext`:

```ruby
Error = Class.new(StandardError)

module HasContext
  attr_reader :context

  def _set_context(context)
    @context = context
  end

  alias_method :original_to_s, :to_s

  def to_s
    if context
      "#{original_to_s} (#{context})"
    else
      original_to_s
    end
  end
end

class CommandError < Error
  attr_reader :command
  include HasContext

  def set_command(command)
    @command = command
  end
end

WrongTypeError = Class.new(CommandError)
foo = WrongTypeError.new("hi there")

puts foo.message
```

This results in `#<WrongTypeError:0x0000000102a48190>` instead of `hi there` because `to_s` is overridden in the `HasContext`.

Reply:
Yes, it's because of you `alias_method :original_to_s, :to_s`, it's done in the module, not when included in the class.

You'd need:

```
def self.included(error_class)
  error_class.alias_method :original_to_s, :to_s
end
```

But a better way to do this is just to call `super` from `to_s`, no need for an alias method.

Reply:
Also note to self, but I need to look at whether `#to_s` is the best way to decorate an error, that has been a number of changes in recent rubies, such as `#detailed_message` etc, and I don't quite remember the whole ordeal.

Reply:
Thanks, I'm not sure why that didn't occur to me.